core/service.go

```diff
@@ -295,21 +303,57 @@ func (s *Service) handleReceivedMessage(msg network.Message) (err error) {
 // announce messages (block announce messages include the header but the full
 // block is required to execute `core_execute_block`).
 func (s *Service) ProcessBlockAnnounceMessage(msg network.Message) error {
+	blockAnnounceMessage, ok := msg.(*network.BlockAnnounceMessage)
+	if !ok {
+		return errors.New("could not cast p2p.Message to BlockAnnounceMessage")
+	}
 
-	// TODO: check if we should send block request message
-
-	// TODO: update message properties and use generated id
-	blockRequest := &network.BlockRequestMessage{
-		ID:            1,
-		RequestedData: 2,
-		StartingBlock: []byte{},
-		EndBlockHash:  optional.NewHash(true, common.Hash{}),
-		Direction:     1,
-		Max:           optional.NewUint32(false, 0),
+	if s.blockState == nil {
+		return errors.New("ProcessBlockAnnounceMessage error: blockState is nil")
 	}
 
-	// send block request message to network service
-	s.msgSend <- blockRequest
+	latestBlockNum := s.blockState.LatestHeader().Number
+	messageBlockNum := blockAnnounceMessage.Number
+
+	// check if we should send block request message
+	if latestBlockNum.Cmp(messageBlockNum) == 1 {
+		// BlockAnnounceMessage
+		blockAnnounce := &network.BlockAnnounceMessage{
+			ParentHash:     blockAnnounceMessage.ParentHash,
+			Number:         messageBlockNum,
+			StateRoot:      blockAnnounceMessage.StateRoot,
+			ExtrinsicsRoot: blockAnnounceMessage.ExtrinsicsRoot,
+			Digest:         blockAnnounceMessage.Digest,
+		}
+
+		// send blockAnnounce message to network service
+		log.Debug("send blockAnnounce message to network service")
+		s.msgSend <- blockAnnounce
```

Comment:
we shouldn't be sending a BlockAnnounce message to the network, we are the ones receiving BlockAnnounce messages from the network. you can remove this whole part and just leave the `else` section

Reply:
correct, no need to send a block announce... the network service will be the one receiving the announce message (and gossiping the message upon receiving it). The only time we send an announce message to the network service is when we produce a new block.

Reply:
right, this is what I understood after our call. but will remove.